plugins/cilium-cni/lib/deletion_queue.go

```diff
@@ -34,73 +46,62 @@ const timeoutSeconds = 10
 // the maximum number of queued deletions allowed, to protect against kubelet insanity
 const maxDeletionFiles = 256
 
-// NewDeletionFallbackClient creates a client that will either issue an EndpointDelete
-// request via the api, *or* queue one in a temporary directory.
-// To prevent race conditions, the logic is:
-// 1. Try and connect to the socket. if that succeeds, done
-// 2. Otherwise, take a shared lock on the delete queue directory
-// 3. Once we get the lock, check to see if the socket now exists
-// 4. If it exists, drop the lock and use the api
-func NewDeletionFallbackClient(logger *slog.Logger) (*DeletionFallbackClient, error) {
-	dc := &DeletionFallbackClient{
-		logger: logger,
-	}
+var (
+	ErrClientFailure = errors.New("client failed")
 
-	// Try and connect (the usual case)
-	err := dc.tryConnect()
-	if err == nil {
-		return dc, nil
-	}
-	dc.logger.Warn(
-		"Failed to connect to agent socket",
-		logfields.Error, err,
-		logfields.Socket, client.DefaultSockPath(),
-	)
+	errClientConnectionFailed = errors.New("client connection failed")
+	errServiceUnavailable     = errors.New("service unavailable")
+)
 
-	// We failed to connect: get the queue lock
-	if err := dc.tryQueueLock(); err != nil {
-		return nil, fmt.Errorf("failed to acquire deletion queue: %w", err)
-	}
+func newCiliumClient(timeout time.Duration) (ciliumClient, error) {
+	return client.NewDefaultClientWithTimeout(timeout)
+}
 
-	// We have the queue lock; try and connect again
-	// just in case the agent finished starting up while we were waiting
-	if err := dc.tryConnect(); err == nil {
-		dc.logger.Info("Successfully connected to API on second try.")
-		// hey, it's back up!
-		dc.lockfile.Unlock()
-		dc.lockfile = nil
-		return dc, nil
-	}
+func wrapClientFailure(err error) error {
+	return fmt.Errorf("%w: %w", ErrClientFailure, err)
+}
+
+// NewDeletionFallbackClient creates a new deletion client.
+func NewDeletionFallbackClient(logger *slog.Logger) *DeletionFallbackClient {
+	return &DeletionFallbackClient{
+		logger: logger,
+
+		deleteQueueDir:      defaults.DeleteQueueDir,
+		deleteQueueLockfile: defaults.DeleteQueueLockfile,
 
-	// We have the lockfile, but no valid client
-	dc.logger.Info("Agent is down, falling back to deletion queue directory")
-	return dc, nil
+		newCiliumClientFn: newCiliumClient,
+	}
 }
 
 func (dc *DeletionFallbackClient) tryConnect() error {
-	c, err := client.NewDefaultClientWithTimeout(timeoutSeconds * time.Second)
+	if dc.cli != nil {
+		return nil
+	}
+
+	c, err := dc.newCiliumClientFn(timeoutSeconds * time.Second)
 	if err != nil {
-		return err
+		return fmt.Errorf("%w: %w", errClientConnectionFailed, err)
```

Comment:
As far as I understand, this is what's ensuring that a timeout error during connect will be handled the same as a service unavailable error in the caller. Solves the concern Marcel was concerned about :+1: 

I dug around in the internals of the client function for a while to try to understand which conditions should or should not be treated like service unavailable, and it was a bit tricky to come to a conclusive understanding about the range of error conditions that might occur as part of the client creation. Maybe this is good enough, and if it turns out that some specific error shouldn't be handled as service unavailable, we can address the issue as it comes up.